server/server_test.go

```diff
@@ -770,6 +774,138 @@ func TestLogSourceIPs(t *testing.T) {
+
+	type tracingSpanJSON struct {
+		OperationName string
+		Tags          map[string]interface{}
+	}
+
+	// writeSpanToBodyHandleFunc provides a way to test tracing by dumping the data
+	// from the requests' currently-active span data into a structured JSON body.
+	//
+	// HTTPGRPCTracer's parent Span tagging cannot be tested with current tracer implementations;
+	// tagging the parentSpan depends on methods only available on the Jaeger tracer, which limits
+	// introspection in tests as jaeger.Tracer can only access the current span, not the parent.
+	//
+	// MockTracer from opentracing/opentracing-go provides visibility into all completed spans for
+	// testing, but does not implement the jaeger operations needed to tag the parent span
+	writeSpanToBodyHandleFunc := func(w http.ResponseWriter, r *http.Request) {
+		span := opentracing.SpanFromContext(r.Context()).(*jaeger.Span)
+		jsonSpan := tracingSpanJSON{
+			OperationName: span.OperationName(),
+			Tags:          span.Tags(),
+		}
+		err := json.NewEncoder(w).Encode(jsonSpan)
+		require.NoError(t, err)
+	}
+
+	helloRouteName := "hello"
+	// explicitly-named routes will be labeled using the provided route name
+	server.HTTP.NewRoute().Name(helloRouteName).Path("/hello").HandlerFunc(writeSpanToBodyHandleFunc)
+
+	helloPathParamRouteTmpl := "/hello/{pathParam}"
+	// unnamed routes will be labeled with their registered path template
+	server.HTTP.HandleFunc(helloPathParamRouteTmpl, writeSpanToBodyHandleFunc)
+
+	// extracted route names or path templates are converted to a Prometheus-compatible label value
+	expectedHelloRouteLabel := middleware.MakeLabelValue(helloRouteName)
+	expectedHelloPathParamRouteLabel := middleware.MakeLabelValue(helloPathParamRouteTmpl)
+
+	go func() {
+		require.NoError(t, server.Run())
+	}()
+
+	helloRouteURL := "http://127.0.0.1/hello"
+	expectedTagsHelloRoute := map[string]interface{}{
+		"component":       "net/http",
+		"span.kind":       "server",
+		"http.url":        helloRouteURL,
+		"http.method":     "GET",
+		"http.route":      expectedHelloRouteLabel,
+		"http.user_agent": "",
+	}
+	helloPathParamRouteURL := "http://127.0.0.1/hello/world"
+	expectedTagsHelloPathParamRoute := map[string]interface{}{
+		"component":       "net/http",
+		"span.kind":       "server",
+		"http.url":        helloPathParamRouteURL,
+		"http.method":     "GET",
+		"http.route":      expectedHelloPathParamRouteLabel,
+		"http.user_agent": "",
+	}
+
+	target := server.GRPCListenAddr()
+	conn, err := grpc.Dial(
+		target.String(),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(4*1024*1024)),
+	)
+	require.NoError(t, err)
+	defer conn.Close()
+	client := httpgrpc.NewHTTPClient(conn)
+
+	// emulateHTTPGRPCPRoxy mimics the usage of the Server type as a load balancing proxy,
+	// wrapping http requests into gRPC requests to utilize gRPC load balancing features
+	emulateHTTPGRPCPRoxy := func(
```

Comment:
[nit] Is this really emulating something? Or is it just a convenience function to reduce the duplication in calling the httpgrpc server for the two test cases?

Reply:
I mean you could say it's more general, but it is emulating how a proxy using this library works.

Otherwise you would not bother creating a normal HTTP request object -
I think when I add the test cases for standard http requests it will become more clear, as then we can show that the same HTTP request object behaves the same whether it's used directly or "proxied" into a gRPC request